IPC::Connection::SyncMessageState does not support multiple runloops

kkinnunen-apple opened the pull request:
#### 452967ca049787c69ec6cd0fd05ff1d138ee8c65
<pre>
IPC::Connection::SyncMessageState does not support multiple runloops
<a href="https://bugs.webkit.org/show_bug.cgi?id=245475">https://bugs.webkit.org/show_bug.cgi?id=245475</a>

Reviewed by Chris Dumez.

Add per-run loop SyncMessageState to IPC::Connection. However,
hard-code it using RunLoop::main() for all connections.
There is no functional difference.

This preparatory change for running Connections in arbitrary
run loops. This feature is needed for IPC from Workers.

* Source/WTF/wtf/RunLoop.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageState::SyncMessageState):
(IPC::Connection::SyncMessageState::WTF_GUARDED_BY_LOCK):
(IPC::Connection::SyncMessageState::get):
(IPC::Connection::SyncMessageState::release):
(IPC::Connection::SyncMessageState::enqueueMatchingMessages):
(IPC::Connection::SyncMessageState::processIncomingMessage):
(IPC::Connection::SyncMessageState::dispatchMessages):
(IPC::Connection::SyncMessageState::dispatchMessagesAndResetDidScheduleDispatchMessagesForConnection):
(IPC::Connection::enqueueMatchingMessagesToMessageReceiveQueue):
(IPC::Connection::open):
(IPC::Connection::platformPrepareForOpen):
(IPC::Connection::invalidate):
(IPC::Connection::waitForMessage):
(IPC::Connection::waitForSyncReply):
(IPC::Connection::processIncomingSyncReply):
(IPC::Connection::processIncomingMessage):
(IPC::Connection::connectionDidClose):
(IPC::Connection::dispatchSyncMessage):
(IPC::Connection::dispatchDidReceiveInvalidMessage):
(IPC::Connection::SyncMessageState::singleton): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::Connection::platformOpen):
(IPC::Connection::open): Deleted.
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::Connection::platformPrepareForOpen):
(IPC::Connection::platformOpen):
(IPC::Connection::open): Deleted.
* Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
(IPC::Connection::platformOpen):
(IPC::Connection::open): Deleted.

Canonical link: <a href="https://commits.webkit.org/255334@main">https://commits.webkit.org/255334@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d5cf284b7cab8b599a381cb846a0ee46997a4d88

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91956 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1200 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101777 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1196 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29642 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84428 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97983 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97614 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/732 "Found 1 new test failure: fast/forms/ios/file-upload-panel-capture.html (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78506 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27689 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82654 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82255 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70725 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36047 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16273 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33785 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17374 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3701 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37662 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40089 "Found 3 new test failures: imported/w3c/web-platform-tests/import-maps/multiple-import-maps/basic.html, imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.html?h264_annexb, imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.worker.html?h264_annexb (failure)") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39546 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36497 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -392,9 +394,11 @@ class Connection : public ThreadSafeRefCounted<Connection, WTF::DestructionThrea
 
     static Lock s_connectionMapLock;
     Client* m_client { nullptr };
+    SyncMessageState* m_syncState { nullptr };
```
I really don't like that this variable is protected by a lock, but there's no way to really know that except by code inspection of the callsites.

I understand that writes only happen from one thread, so we don't need to use the lock for those locations, but that's going to be somewhat hard for a future reader to know without careful analysis.

This is only going to get harder once we try to access data on the SyncMessageState (like the SerialFunctionDispatcher), since there will be even more callsites where it's hard to know what's safe to access, and how it can be guaranteed to be alive for.

I'm really a huge fan of clang's thread safety annotations (or DataMutex :)), since then the compiler can immediately tell you if you're adding a bug (for this subset of bugs). Obviously that means we have to lock on all threads, but my opinion is that this is a really good tradeoff.

If the performance cost of the extra locking really is a problem, then it seems possible to come up with a 'single writer, multiple readers' pattern (maybe a variant of DataMutex?) which makes it harder to get wrong. Can't determine safety statically, but we could at least have named accessors (like `accessFromWriterThread`) which force you to declare your intent.

kkinnunen-apple replied:
> I really don't like that this variable is protected by a lock, but there's no way to really know that except by code inspection of the callsites.

Yeah, it's not ideal, though it is an established pattern in WebKit code wrt "variable is mostly accessed from the maintop".

> If the performance cost of the extra locking really is a problem

It's mostly not a perf problem. It's a consistency problem. You look at the code and you see that clearly the variable is only ever modified from the main thread. Yet, you see all main thread read accesses lock the read. What's up with that.

You can contrast:
- No explanation of the locking: leaves the reader questioning.
- Excess locking: leaves the code inconsistent and leaves the reader questioning.

I added the static analysis support for the run-time properties.  How does the patch look now?

mattwoodrow replied:
I'm not sure I entirely understand the new static analysis, isn't the LockAssertion object always available (since it's a static const), so the static analysis can't catch failing cases? There's a lot of extra code added, and it's hard (for me) to understand what constraints it's able to enforce.

I guess I just feel like excess locking is less likely to leave the reader questioning. The variable would be locked for every access, and the compiler would fail if you ever wrote code that didn't do that. It's indeed unnecessary in some cases, but it's at least a very clearly enforced constraint, despite being a superset of the minimum 'safe' constraints.

I'm not yet a WebKit reviewer, so it might be best for @cdumez or someone to look at this time, since it'll require their sign off regardless.

kkinnunen-apple replied:
> I'm not sure I entirely understand the new static analysis, isn't the LockAssertion object always available (since it's a static const), so the static analysis can't catch failing cases? 

Static analysis of run-time properties cannot catch failing nor non-failing cases. That's why it's a run-time property, if if was not, it would be a compile-time property.  :) 

However, there is a test of the system proves that this catches non-trivial negative cases.  The catch is that it's a runtime error, e.g. an ASSERT on lock incorrectly not held.

It also catches trivial negative cases, e.g. missing assertions about the locks.

The system added here is the same as with asserting that code is run in particular thread:

```
struct {
   int n; 
   int m WTF_GUARDED_BY(m_workQueue);
} z;

void myFeatureThatIsRunOnWorkQueue() {
    assertIsCurrent(m_workQueue);
    z.m = z.n +1;
}
```

The compiler utterly just cannot know that the code is run in work queue.
What the compiler thinks:

- Programmer tells me that the code is run on the work queue.
- Accepting this as a fact, `z.m` can be assigned to since it can be only assigned when m_workQueue is current.

What the programmer thinks and tells the computer to do:
- I know the code is run only on the work queue.
- If during runtime we enter here, please ASSERT if we are not on the work queue.
- If I don't see the assertions, I'm confident I'm not lying to the compiler that the code is run on the work queue.

What the computer does during runtime:
- Check that the current work queue is the passed in work queue. ASSERT if not.

Similarly:
```
const LockAssertion& myTuesdaysLock() { if (isTuesday()) return LockAssertion::shared; return LockAssertion::unlocked; }
struct {
   int n WTF_ GUARDED_BY(myTuesdaysLock());
   int m;
} z;

void myFeatureThatIsRunOnTuesdays() {
   assertIsShared(myTuesdaysLock());
   z.m = z.n + 1;
}
```

Here the compiler thinks:
- Programmer tells me that the lock returned by `myTuesdaysLock()` is held as shared.
- Based on this, the assignment is correct to read the variable `z.n`, since that variable can be only read when `myTuesdaysLock()` is held as shared or exclusive.

What the programmer thinks and tells the computer to do:
- I know based on how I wrote the code that this code is only run on Tuesdays.
- Computer, please ASSERT if we are running the code on any other day than Tuesday.
- If I don't see the assertions, I'm confident I'm not lying to the compiler that the code is run on Tuesdays only.

What the computer does:
- Check that myTuesdaysLock() returns shared. ASSERT if not.


> There's a lot of extra code added, and it's hard (for me) to understand what constraints it's able to enforce.

It's able to enforce everything, but it's a bit cumbersome since it does not understand the underlying locks, rather it must rely on the explicit assertions (similar to the is current case). It is also vulnerable to the incorrectly specified lock conditions (like everything?).

mattwoodrow commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -431,13 +453,37 @@ void Connection::setDidCloseOnConnectionWorkQueueCallback(DidCloseOnConnectionWo
+        Locker locker { m_incomingMessagesLock };
+        syncState = std::exchange(m_syncState, nullptr);
+    }
+    syncState->release();
```
I feel sort of similar here, with the use of a manual `release`. I think we only need to rename this to `deref`, and then `RefPtr<>` will work.

It's only a single callsite (and is unlikely to ever be more than that), but we have a project-wide convention of avoiding manual refcounting and relying on the auto wrapper objects, so I don't see a reason to deviate here.

kkinnunen-apple replied:
But it's not a RefPtr if it doesn't really support RefPtr. This is why I would not want to shoehorn it using it.
I don't think lack of T::ref() and still making the RefPtr work is by design, it's just an emergent feature.

It's relatively simple use in here: the destructor asserts that the code in question has run.

I added a bit of convolution but can fix it up in further patch when moving SyncMessageState to its own file to simplify the implementations. Would this be ok for you?

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -128,23 +134,40 @@ class Connection::SyncMessageState {
     Deque<ConnectionAndIncomingMessage> m_messagesToDispatchWhileWaitingForSyncReply WTF_GUARDED_BY_LOCK(m_lock);
+
+    RunLoop& m_runLoop;
+    int m_clients WTF_GUARDED_BY_LOCK(syncMessageStateMapLock) { 0 };
```
Why are we using an int and not an unsigned? Looks a little odd.

kkinnunen-apple replied:
I think in C++ the idea is that all other aspects being equal, int is the "more optimal" in the way that it communicates to the compiler that the value  value cannot overflow.

underflow in unsigned -- if possible, it would be my error
overflow in unsigned -- if possible, it would be my error

going negative in int -- if possible, it would be my error
overflow in int -- if possible, it would be my error

I can of course change it to unsigned..

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -102,13 +100,21 @@ class Connection::SyncMessageState {
+    static Lock syncMessageStateMapLock;
+    static HashMap<RunLoop*, SyncMessageState*>& syncMessageStateMap() WTF_REQUIRES_LOCK(syncMessageStateMapLock)
+    {
+        static NeverDestroyed<HashMap<RunLoop*, SyncMessageState*>> map;
```
can we use a std::unique_ptr<> for the value instead of a raw pointer? We frown upon explicit new / delete in WebKit code.

kkinnunen-apple replied:
Not here, destruction is not available to unique_ptr because there is no unique owner for the instance. The intention is that nobody can call the destructor except the deleter, because in no other way the deletion can be correct.

It's a reference counted instance with non-copyable references. SyncMessageState::get() is the factory for the references. I'll make sure the ::get() will already return the unique_ptr, though

Each reference has an unique owner. The unique_ptr is used to hold each reference via custom unique_ptr.

Using the unique_ptr here is more of an artefact for the class being inside Connection.h . If the SyncMessageState would be in its own header, it could be called something like SyncMessageStateHolder as there would be more room to type up the ceremony in the individual header file.

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -128,23 +134,40 @@ class Connection::SyncMessageState {
+            return;
+        Connection::SyncMessageState::syncMessageStateMap().remove(&instance->m_runLoop);
+    }
+    delete instance;
```
Ideally we'd use a unique_ptr<> with makeUnique<> and avoid explicit new / delete.

kkinnunen-apple replied:
This is the unique_ptr<> destroy operator you're commenting on :)

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -431,13 +455,35 @@ void Connection::setDidCloseOnConnectionWorkQueueCallback(DidCloseOnConnectionWo
     m_client = nullptr;
+    [this] {
+        Locker locker { m_incomingMessagesLock };
+        return WTFMove(m_syncState);
```
Why `return WTFMove(m_syncState);` and not simply `m_syncState = nullptr;`?

kkinnunen-apple replied:
By default running minimal amount of code under lock. In this case, it avoids running the sync state destructor under the lock. Which should never matter, but just trying to adhere to the the default policy.

cdumez reviewed (approved):
LGTM with comments.

kkinnunen-apple committed:
IPC::Connection::SyncMessageState does not support multiple runloops
https://bugs.webkit.org/show_bug.cgi?id=245475

Reviewed by Chris Dumez.

Add per-run loop SyncMessageState to IPC::Connection. However,
hard-code it using RunLoop::main() for all connections.
There is no functional difference.

This preparatory change for running Connections in arbitrary
run loops. This feature is needed for IPC from Workers.

* Source/WTF/wtf/RunLoop.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageState::SyncMessageState):
(IPC::Connection::SyncMessageState::WTF_GUARDED_BY_LOCK):
(IPC::Connection::SyncMessageState::get):
(IPC::Connection::SyncMessageState::release):
(IPC::Connection::SyncMessageState::enqueueMatchingMessages):
(IPC::Connection::SyncMessageState::processIncomingMessage):
(IPC::Connection::SyncMessageState::dispatchMessages):
(IPC::Connection::SyncMessageState::dispatchMessagesAndResetDidScheduleDispatchMessagesForConnection):
(IPC::Connection::enqueueMatchingMessagesToMessageReceiveQueue):
(IPC::Connection::open):
(IPC::Connection::platformPrepareForOpen):
(IPC::Connection::invalidate):
(IPC::Connection::waitForMessage):
(IPC::Connection::waitForSyncReply):
(IPC::Connection::processIncomingSyncReply):
(IPC::Connection::processIncomingMessage):
(IPC::Connection::connectionDidClose):
(IPC::Connection::dispatchSyncMessage):
(IPC::Connection::dispatchDidReceiveInvalidMessage):
(IPC::Connection::SyncMessageState::singleton): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::Connection::platformOpen):
(IPC::Connection::open): Deleted.
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::Connection::platformPrepareForOpen):
(IPC::Connection::platformOpen):
(IPC::Connection::open): Deleted.
* Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
(IPC::Connection::platformOpen):
(IPC::Connection::open): Deleted.

Canonical link: https://commits.webkit.org/255334@main
